streaming/src/main/scala/org/apache/spark/streaming/dstream/SocketInputDStream.scala

```diff
@@ -52,46 +52,60 @@ class SocketReceiver[T: ClassTag](
     storageLevel: StorageLevel
   ) extends Receiver[T](storageLevel) with Logging {
 
+  private var socket: Socket = _
+
   def onStart() {
-    // Start the thread that receives data over a connection
-    new Thread("Socket Receiver") {
-      setDaemon(true)
-      override def run() { receive() }
-    }.start()
+
+    try {
+      logInfo(s"Connecting to $host:$port")
+      socket = new Socket(host, port)
+      logInfo(s"Connected to $host:$port")
+
+      // Start the thread that receives data over a connection
+      new Thread("Socket Receiver") {
+        setDaemon(true)
+        override def run() { receive() }
+      }.start()
+    } catch {
+      case e: ConnectException =>
+        restart(s"Error connecting to $host:$port", e)
+      case NonFatal(e) =>
```

Comment:
"This is now in the wrong place. It was correct where it was."
Sorry, I didn't catch this comment. I thought you means the "new Thread" block so I moved the "new Thead" block outside the try/catch block,  It will looks like cause NPE.  could you give more information about which block is now in the wrong block?